registry/handlers/referrers.go

```diff
@@ -0,0 +1,366 @@
+		if err != nil {
+			h.Errors = append(h.Errors, fmt.Errorf("nextToken base64 decoding failed")) // need to check the v2 error code. check with origin implementation plz
+			return
+		}
+		nextTokenList := strings.Split(string(nextTokenDecoded), ",")
+		for _, token := range nextTokenList {
+			_, err := digest.Parse(token)
+			if err != nil {
+				h.Errors = append(h.Errors, fmt.Errorf("nextToken parsing failed")) // same as above.
+				return
+			}
+			// store nextToken digest in a map for quick access
+			nextTokenMap[token] = token
+		}
+	}
+
+	if h.Digest == "" {
+		h.Errors = append(h.Errors, v2.ErrorCodeManifestUnknown.WithDetail("digest not specified"))
+		return
+	}
+
+	referrers, err := h.generateSortedReferrers(h, h.Digest, artifactType)
+	if err != nil {
+		if _, ok := err.(distribution.ErrManifestUnknownRevision); ok {
+			h.Errors = append(h.Errors, v2.ErrorCodeManifestUnknown.WithDetail(err))
+		} else {
+			h.Errors = append(h.Errors, errcode.ErrorCodeUnknown.WithDetail(err))
+		}
+		return
+	}
+
+	if referrers == nil {
+		referrers = []v1.Descriptor{}
+	}
+
+	// only consider pagination if # of referrers is greater than page size
+	if len(referrers) > nPage {
+		startIndex := 0
+		if len(nextTokenMap) > 0 {
+			for i, ref := range referrers {
+				// check if ref matches a digest in nextToken list
+				if _, ok := nextTokenMap[ref.Digest.String()]; ok {
+					// set the starting index to the largest index
+					if (i + 1) > startIndex {
+						startIndex = i + 1
+					}
+				}
+			}
+		}
+
+		// only applicable if last item provided as nextToken
+		if startIndex == len(referrers) {
+			referrers = []v1.Descriptor{}
+		}
+
+		// if there's only 1 page of results left
+		if len(referrers)-startIndex <= nPage {
+			referrers = referrers[startIndex:]
+		} else {
+			referrers = referrers[startIndex:(startIndex + nPage)]
+			// nextToken is a base64 encoded comma-separated string of the digests of the last three referrers in the response
+			var nextDgsts []string
+			for i := nPage - 1; i >= nPage-minPageSize; i-- {
+				nextDgsts = append(nextDgsts, referrers[i].Digest.String())
+			}
+			// if n was not provided in page, set nPage to a value so link header knows not to include n
+			if nParseError != nil {
+				nPage = -1
+			}
+			// add the Link Header
+			w.Header().Set("Link", generateLinkHeader(h.Repository.Named().Name(), h.Digest.String(), artifactType, nextDgsts, nPage))
+		}
+	}
+
+	response := ReferrersAPIResponse{
+		SchemaVersion: 2,
+		MediaType:     v1.MediaTypeImageIndex,
+		Manifests:     referrers,
+	}
+
+	w.Header().Set("Content-Type", v1.MediaTypeImageIndex)
+	enc := json.NewEncoder(w)
+	if err = enc.Encode(response); err != nil {
+		h.Errors = append(h.Errors, errcode.ErrorCodeUnknown.WithDetail(err))
+		return
+	}
+}
+
+func (h *referrersHandler) generateSortedReferrers(ctx context.Context, revision digest.Digest, artifactType string) ([]v1.Descriptor, error) {
```

Comment:
Does OCI Referrer Spec specifies sorting? If so, please point out the spec reference.

Reply:
Removed sorting